libtelnet.c

```diff
@@ -1121,6 +1137,47 @@ static void _process(telnet_t *telnet, const char *buffer, size_t size) {
 	}
 }
 
+/* perform NVT end-of-line translation for received text */
+size_t telnet_translate_eol(telnet_t *telnet, char *buffer, size_t size,
+		int *split) {
+	size_t i = 0, l = 0;
+
+	/* no translation if in BINARY mode */
+	if (telnet->flags & TELNET_FLAG_RECEIVE_BINARY)
+		return size;
+
+	/* fix up translation if split across buffer boundary */
+	if (*split && size > 0) {
+		/* CR-NUL translates to \r which was skipped at end of previous */
+		/* buffer, while CR-LF translates to \n, so just leave it */
+		if (buffer[i] == '\0') {
+			buffer[i] = '\r';
+		}
+		++i;
+	}
+
+	for ( ; i < size-1; ++i, ++l) {
+		buffer[l] = buffer[i];
+		if (buffer[i] == '\r') {
+			if (buffer[i+1] == '\n') {
+				/* CR-LF translates to \n */
+				buffer[i++] = '\n';
+			} else if (buffer[i+1] == '\0') {
+				/* CR-NUL translates to \r */
+				++i;
+			}
+		}
+	}
+	if (size > 0) {
+		if (buffer[i] == '\r') {
+			--l;
+			*split = 1;
+		} else
+			*split = 0;
+	}
+	return size - (i - l);
```

Comment:
If I can't tell if then return value is correct with a glance (and I can't) it probably needs some refactoring, better variable names, or so on. One approach may be to just modify `size` or a copy there-of rather than trying to recompute it.